src/hypothesis/stateful.py

```diff
@@ -354,13 +353,59 @@ def precondition_wrapper(*args, **kwargs):
             setattr(precondition_wrapper, PRECONDITION_MARKER, precond)
         else:
             new_rule = Rule(targets=rule.targets, arguments=rule.arguments,
-                            function=rule.function, precondition=precond,
-                            parent_rule=rule.parent_rule)
+                            function=rule.function, precondition=precond)
             setattr(precondition_wrapper, RULE_MARKER, new_rule)
+
+        invariant = getattr(f, INVARIANT_MARKER, None)
+        if invariant is not None:
+            new_invariant = Invariant(function=invariant.function,
+                                      precondition=precond)
+            setattr(precondition_wrapper, INVARIANT_MARKER, new_invariant)
+
         return precondition_wrapper
     return decorator
 
 
+Invariant = namedtuple(
+    u'Invariant',
+    (u'function', u'precondition')
+)
+
+
+def invariant():
```

Comment:
I can imagine wanting to check invariants of bundles or the like.

Most of the code I've written just has `@rule()` without any arguments, so I made `@invaraint()` behave the same way. Given that both `rule` and `precondition` need to be called, I think it makes sense for `invariant` to need to be called as well.